lib/matplotlib/figure.py

```diff
@@ -1374,14 +1394,68 @@ def legend(self, handles, labels, *args, **kwargs):
         columnspacing      the spacing between columns
         ================   ====================================================
 
-        .. Note:: Not all kinds of artist are supported by the legend.
-                  See LINK (FIXME) for details.
+        .. Note:: Not all kinds of artist are supported by the legend command.
+                  See :ref:`plotting-guide-legend` for details.
 
         **Example:**
 
         .. plot:: mpl_examples/pylab_examples/figlegend_demo.py
         """
-        l = Legend(self, handles, labels, *args, **kwargs)
+
+        if len(args) == 0:
+            handles = []
+            labels = []
+
+            def in_handles(h, l):
+                for f_h, f_l in zip(handles, labels):
+                    if f_l != l:
+                        continue
+                    if type(f_h) != type(h):
+                        continue
+
+                    try:
+                        if f_h.get_color() != h.get_color():
+                            continue
+                    except AttributeError:
+                        pass
+                    try:
+                        if f_h.get_facecolor() != h.get_facecolor():
+                            continue
+                    except AttributeError:
+                        pass
+                    try:
+                        if f_h.get_edgecolor() != h.get_edgecolor():
+                            continue
+                    except AttributeError:
+                        pass
+
+                    return True
+                return False
+
+            for ax in self.axes:
+                ax_handles, ax_labels = ax.get_legend_handles_labels()
+                for h, l in zip(ax_handles, ax_labels):
+                    if not in_handles(h, l):
+                        handles.append(h)
+                        labels.append(l)
+            if len(handles) == 0:
+                warnings.warn("No labeled objects found. "
+                              "Use label='...' kwarg on individual plots.")
+                return None
+
+        elif len(args) == 2:
+            # LINES, LABELS
+            handles, labels = args
+
+        elif len(args) == 3:
+            # LINES, LABELS, LOC
+            handles, labels, loc = args
+            kwargs['loc'] = loc
+
+        else:
+            raise TypeError('Invalid arguments to legend')
```

Comment:
Do you think we could have an error message more specific? It's quite obscure right now.
I like better "Invalide number of arguments to legend. Possible arguments are (etc)"